src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,176 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node first;
+    private Node last;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> addNode;
+        if (size == 0) {
+            addNode = new Node<>(null, value, null);
+            first = addNode;
+
+        } else {
+            addNode = new Node<>(last, value, null);
+            last.next = addNode;
+        }
+        last = addNode;
+        size++;
     }
 
-    @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Invalid index: " + index);
+        }
+
+        if (index == size) {
+            Node<T> oldLast = last;
+            Node<T> newNode = new Node<>(oldLast, value, null);
+            last = newNode;
+            if (oldLast == null) {
+                first = newNode;
+            } else {
+                oldLast.next = newNode;
+            }
+        } else if (index == 0) {
+            Node<T> oldFirst = first;
+            Node<T> newNode = new Node<>(null, value, oldFirst);
+            first = newNode;
+            if (oldFirst == null) {
+                last = newNode;
+            } else {
+                oldFirst.prev = newNode;
+            }
+        } else {
+            Node<T> currentNode = findNodeByIndex(index);
+            Node<T> prevNode = currentNode.prev;
+            Node<T> newNode = new Node<>(prevNode, value, currentNode);
+            prevNode.next = newNode;
+            currentNode.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list == null) {
+            throw new RuntimeException("Can't add list because it is null");
+        }
+        if (list.size() == 0) {
+            return;
+        }
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexValidation(index);
+        Node<T> currentNode = findNodeByIndex(index);
+        return currentNode.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (value == null) {
+            throw new IllegalArgumentException("Cannot set the value because it is null");
+        }
+        indexValidation(index);
+        Node<T> currentNode = findNodeByIndex(index);
```

Comment:
Move index validation inside the "find by index" function since you always use them together.